lib/application.js

```diff
@@ -383,6 +383,53 @@ app.set = function set(setting, val) {
   return this;
 };
 
+/**
+ * Extend `setting` with `val`, or return `setting`'s value.
+ *
+ *    app.set('foo', { 'bar' : 1 })
+ *    app.expand('foo', { 'baz' : 1 });
+ *    app.expand('foo');
+ *    // => { "bar" : 1, "baz" : 1 }
+ *
+ * Mounted servers inherit their parent server's settings.
+ *
+ * @param {String} setting
+ * @param {*} [val]
+ * @return {Server} for chaining
+ * @public
+ */
+
+app.extend = function(setting, val) {
+  if (arguments.length === 1) {
+    // app.get(setting)
+    return this.settings[setting];
+  }
+
+  if (typeof val === 'undefined') {
+    return this;
+  } else if (val === null || typeof val !== 'object') {
+    throw new Error('value needs to be an object');
+  }
+
+  debug('extend "%s" with %o', setting, val);
+
+  var currentSetting = this.settings[setting];
+
+  // no object yet, set value
+  if (currentSetting === null || typeof currentSetting !== 'object') {
+    this.settings[setting] = val;
+    return this;
+  }
+
+  // extend from the intial object type
+  var newObject = Array.isArray(this.settings[setting]) ? [] : {};
```

Comment:
I believe that by adding this `isArray` check you loose support for node `0.12`, making this unable to release on `4.x`.  Also using `Object.assign` on an array is VERY miss leading to consumers.  If I were you I would make this method ONLY work on plain objects and throw like you did above on line 411.